addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRMeterDiscoveryService.java

```diff
@@ -62,7 +76,90 @@ public void telegramReceived(P1Telegram telegram) {
+        validateConfiguredMeters(dsmrBridgeHandler.getThing().getThings(),
+                detectedMeters.getKey().stream().map(md -> md.getMeterType()).collect(Collectors.toSet()));
+        detectedMeters.getKey().forEach(m -> meterDiscovered(m, dsmrBridgeHandler.getThing().getUID()));
+    }
+
+    protected void verifyUnregisteredCosemObjects(P1Telegram telegram,
+            Map<CosemObjectType, CosemObject> undetectedCosemObjects) {
+        if (!undetectedCosemObjects.isEmpty()) {
+            if (undetectedCosemObjects.entrySet().stream()
+                    .anyMatch(e -> e.getKey() == CosemObjectType.METER_EQUIPMENT_IDENTIFIER
+                            && e.getValue().getCosemValues().entrySet().stream().anyMatch(
+                                    cv -> cv.getValue() instanceof StringType && cv.getValue().toString().isEmpty()))) {
+                // Unregistered meter detected. log to the user.
+                reportUnregisteredMeters();
+            } else {
+                reportUnrecognizedCosemObjects(undetectedCosemObjects);
+                logger.info(
+                        "There are some unrecognized values, which means some meters might not be detected. Not all values are recognized. Please report your raw data as example:",
+                        telegram.getRawTelegram());
+            }
+        }
+    }
+
+    /**
+     * Called when Unrecognized cosem objects where found. This can be a bug or a new meter not yet supported.
+     *
+     * @param unidentifiedCosemObjects Map with the unrecognized.
+     */
+    protected void reportUnrecognizedCosemObjects(Map<CosemObjectType, CosemObject> unidentifiedCosemObjects) {
+        unidentifiedCosemObjects
+                .forEach((k, v) -> logger.debug("Unrecognized cosem object '{}' found in the data: {}", k, v));
+    }
+
+    /**
+     * Called when a meter equipment identifier is found that has an empty value. This
+     */
+    protected void reportUnregisteredMeters() {
+        logger.info(
+                "An unregistered meter has been found. Probably a new meter. Retry discovery once the meter is registered with the energy provider.");
+    }
+
+    /**
+     * Validates if the meters configured by the user match with what is detected in the telegram. Some meters are a
+     * subset of other meters and therefore an invalid configured meter does work, but not all available data is
+     * available to the user.
+     *
+     * @param things The list of configured things
+     * @param configuredMeterTypes The set of meters detected in the telegram
+     */
+    private void validateConfiguredMeters(List<Thing> things, Set<DSMRMeterType> configuredMeterTypes) {
+        final Set<DSMRMeterType> configuredMeters = things.stream()
+                .filter(t -> t.getHandler() instanceof DSMRMeterHandler
+                        && ((DSMRMeterHandler) t.getHandler()).getMeterDescriptor() != null)
```

Comment:
Personally I would prefer every operation in a stream to be a single thing, maybe the following would work and be even more comprehensible:

```
things.stream()
                .filter(DSMRMeterHandler.class::isInstance)
                .map(DSMRMeterHandler.class::cast)
                .filter(h -> h.getMeterDescriptor() != null)
                .map(h -> h.getHandler().getMeterDescriptor().getMeterType())
                .collect(Collectors.toSet());

Reply:
I've updated the code with your suggestion (with some small changes).

Reply:
I like yours even better 👍